thrift.js

```diff
@@ -114,48 +158,139 @@ function Thrift(options) {
+            return cb(err);
+        }
+        model.idls[filename] = source;
+        model._asyncParseIncludedModules(filename, allowIncludeAlias, cb);
+    })
+}
+
+Thrift.prototype._asyncParseIncludedModules = function _asyncParseIncludedModules(filename, allowIncludeAlias, cb) {
+    var model = this;
+    var dirname = path.dirname(filename);
+    var defs;
+    try {
+        if (!model.asts[filename]) {
+            model.asts[filename] = idl.parse(model.idls[filename]);
+        }
+        defs = model.asts[filename].headers.concat(model.asts[filename].definitions);
+        model._checkIncludedModules(dirname, defs, allowIncludeAlias);
+    } catch (err) {
+        return cb(err);
+    }
+    asyncEach(defs, function (def, handleCb) {
+        if (def.type !== 'Include') {
+            return handleCb(undefined);
+        }
+        var includeFilename = path.join(dirname, def.id);
+        model._asyncParse(includeFilename, true, function (err) {
+            handleCb(err);
+        });
+    }, cb);
+}
+
+Thrift.prototype._parse = function _parse(filename, allowIncludeAlias) {
+    if (this.parsed[filename]) {
+        return ;
+    }
+    this.parsed[filename] = true;
+
+    if (!this.idls[filename] && !this.asts[filename]) {
         /* eslint-disable max-len */
-        assert.ok(this.fs, this.filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
-        assert.ok(this.filename, 'Thrift must be constructed with a options.entryPoint');
+        assert.ok(this.fs, filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
         /* eslint-enable max-len */
-        this.filename = path.resolve(this.filename);
-        source = this.fs.readFileSync(this.filename, 'ascii');
-        this.idls[this.filename] = source;
+        this.idls[filename] = this.fs.readFileSync(path.resolve(filename), 'ascii');
     }
 
-    this.exception = null;
+    if (!this.asts[filename]) {
+        this.asts[filename] = idl.parse(this.idls[filename]);
+    }
+
+    var dirname = path.dirname(filename);
+    var defs = this.asts[filename].headers.concat(this.asts[filename].definitions);
+    this._checkIncludedModules(dirname, defs, allowIncludeAlias);
+    for (var index = 0; index < defs.length; index++) {
+        var def = defs[index];
+        if (def.type !== 'Include') {
+            continue ;
+        }
+        var includeFilename = path.join(dirname, def.id);
+        this._parse(includeFilename, true);
+    }
+}
+
+Thrift.prototype._checkIncludedModules = function _checkIncludedModules(dirname, defs, allowIncludeAlias) {
+    for (var index = 0; index < defs.length; index++) {
+        var def = defs[index];
+        if (def.type !== 'Include') {
+            continue ;
+        }
+        if (def.id.lastIndexOf('/', 0) === 0) {
+            throw Error('Include path string must not be an absolute path');
```

Comment:
Can you verify that there is a test that ensures that this behavior is unchanged by the move?

Reply:
This test ensures that: https://github.com/dbousque/thriftrw-node/blob/async-load/test/include.js#L133.